src/Daemon/Daemon.cpp

```diff
@@ -460,6 +468,64 @@ namespace usbguard
   {
   }
 
+  void Daemon::daemonize(const std::string &pid_file)
+  {
+      pid_t pid = 0;
+      int fd;
+
+      pid = fork();
+      if (pid < 0) {
+          USBGUARD_LOG(Error) << "Fork failed with " << pid << ", exiting.";
+          exit(EXIT_FAILURE);
+      }
+      if (pid > 0) {
+          exit(EXIT_SUCCESS);
+      }
+
+      /* Now we are forked */
+      if (setsid() < 0) {
+          exit(EXIT_FAILURE);
+      }
+      signal(SIGCHLD, SIG_IGN);
+
+
+      pid = fork();
+      if (pid < 0) {
+          USBGUARD_LOG(Error) << "Fork failed with " << pid << ", exiting.";
+          exit(EXIT_FAILURE);
+      }
+      if (pid > 0) {
+          exit(EXIT_SUCCESS);
+      }
+
+      /* Now we are forked 2nd time */
+      umask(0);
+      chdir("/");
+      for (fd = sysconf(_SC_OPEN_MAX); fd > 0; fd--) {
+              close(fd);
+      }
+      stdin = fopen("/dev/null", "r");
+      stdout = fopen("/dev/null", "w+");
+      stderr = fopen("/dev/null", "w+");
+
+      pid_fd = open(pid_file.c_str(), O_RDWR|O_CREAT, 0640);
+      if (pid_fd < 0) {
+          USBGUARD_LOG(Error) << "Could not open PID file, exiting.";
+          exit(EXIT_FAILURE);
+      }
+      if (lockf(pid_fd, F_TLOCK, 0) < 0) {
+          USBGUARD_LOG(Error) << "Could not lock PID file, exiting.";
+          exit(EXIT_FAILURE);
+      }
+      pid = getpid();
+      char pid_str[16]; /* enough for maximum PID in string form */
+      int len = snprintf(pid_str, 16, "%lld", static_cast<long long int>(pid));
+      if (len < 1) {
+          exit(EXIT_FAILURE);
+      }
+      write(pid_fd, pid_str, len);
```

Comment:
If the open() call succeeds but the write() fails for some reason, shouldn't we try to remove/unlink() the file, so that we try to not leave an invalid PID file left behind?